src/14_command_buffers.rs

```diff
@@ -0,0 +1,632 @@
+}
+
+fn create_surface(
+    instance: &back::Instance,
+    window: &winit::Window,
+) -> <back::Backend as hal::Backend>::Surface {
+    instance.create_surface(window)
+}
+
+fn create_device_with_graphics_queues(
+    adapter: &mut hal::Adapter<back::Backend>,
+    surface: &<back::Backend as hal::Backend>::Surface,
+) -> (
+    <back::Backend as hal::Backend>::Device,
+    Vec<hal::queue::CommandQueue<back::Backend, hal::Graphics>>,
+    hal::queue::QueueType,
+    hal::queue::family::QueueFamilyId,
+) {
+    let family = adapter
+        .queue_families
+        .iter()
+        .find(|family| {
+            hal::Graphics::supported_by(family.queue_type())
+                && family.max_queues() > 0
+                && surface.supports_queue_family(family)
+        }).expect("Could not find a queue family supporting graphics.");
+
+    let priorities = vec![1.0; 1];
+
+    let families = [(family, priorities.as_slice())];
+
+    let hal::Gpu { device, mut queues } = adapter
+        .physical_device
+        .open(&families)
+        .expect("Could not create device.");
+
+    let mut queue_group = queues
+        .take::<hal::Graphics>(family.id())
+        .expect("Could not take ownership of relevant queue group.");
+
+    let command_queues: Vec<_> = queue_group.queues.drain(..1).collect();
+
+    (device, command_queues, family.queue_type(), family.id())
+}
+
+fn find_queue_families(adapter: &hal::Adapter<back::Backend>) -> QueueFamilyIds {
+    let mut queue_family_ids = QueueFamilyIds::default();
+
+    for queue_family in &adapter.queue_families {
+        if queue_family.max_queues() > 0 && queue_family.supports_graphics() {
+            queue_family_ids.graphics_family = Some(queue_family.id());
+        }
+
+        if queue_family_ids.is_complete() {
+            break;
+        }
+    }
+
+    queue_family_ids
+}
+
+fn is_adapter_suitable(adapter: &hal::Adapter<back::Backend>) -> bool {
+    find_queue_families(adapter).is_complete()
+}
+
+fn pick_adapter(instance: &back::Instance) -> hal::Adapter<back::Backend> {
+    let adapters = instance.enumerate_adapters();
+    for adapter in adapters {
+        if is_adapter_suitable(&adapter) {
+            return adapter;
+        }
+    }
+    panic!("No suitable adapter");
+}
+
+#[derive(Default)]
+struct QueueFamilyIds {
+    graphics_family: Option<hal::queue::QueueFamilyId>,
+}
+
+impl QueueFamilyIds {
+    fn is_complete(&self) -> bool {
+        self.graphics_family.is_some()
+    }
+}
+
+fn init_window() -> (winit::Window, winit::EventsLoop) {
+    let events_loop = winit::EventsLoop::new();
+    let window_builder = winit::WindowBuilder::new()
+        .with_dimensions(winit::dpi::LogicalSize::new(1024., 768.))
+        .with_title(WINDOW_NAME.to_string());
+    let window = window_builder.build(&events_loop).unwrap();
+    (window, events_loop)
+}
+
+fn create_instance() -> back::Instance {
+    back::Instance::create(WINDOW_NAME, 1)
+}
+
+fn clean_up(
```

Comment:
Could we keep the ordering of the functions the same (`clean_up` after `init_hal`) to make the diffs easier?